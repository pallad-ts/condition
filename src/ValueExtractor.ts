import * as is from 'predicates';
import {DENORMALIZE, NORMALIZE} from "./global";

export abstract class ValueExtractor<T> {
    abstract extract(value: T): Promise<any> | any;

    abstract toString(): string;
}

export namespace ValueExtractor {

    export class Property<T> extends ValueExtractor<T> {
        readonly path: Array<string | symbol>;

        constructor(...path: Property.Path) {
            if (path.length <= 0) {
                throw new Error('Property path must be longer than 0');
            }

            super();
            this.path = path;
            Object.freeze(this);
        }

        static create(...path: Property.Path) {
            return new Property(...path);
        }

        async extract(value: T): Promise<any> {
            let currentValue: any = value;
            for (const part of this.path) {
                if (is.object(currentValue)) {
                    currentValue = await currentValue[part];
                } else {
                    return undefined;
                }
            }
            return currentValue;
        }

        toString() {
            return `property "${this.path.join('.')}"`;
        }

        [NORMALIZE]() {
            return this.path;
        }

        static [DENORMALIZE](data: string[]) {
            return new Property(...data);
        }
    }

    export namespace Property {
        export type Path = Part[];
        export type Part = symbol | string;
    }
}
