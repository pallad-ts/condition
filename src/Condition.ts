import * as is from 'predicates';
import {Predicate} from "predicates/types";
import {DENORMALIZE, NORMALIZE} from "./global";
import {ValueExtractor} from "./ValueExtractor";

export abstract class Condition<T = any> {
    abstract isSatisfied(value: T): Promise<boolean> | boolean;

    abstract toString(): string;
}

export function createConditionClass<TPredicate extends (...args: any[]) => Predicate<any>>(
    predicateFactory: TPredicate,
    descFactory?: (...args: Parameters<TPredicate>) => string,
    normalizeArgs?: (...args: Parameters<TPredicate>) => any,
    denormalizeArgs?: (...args: any[]) => Parameters<TPredicate>
): { new<T>(...args: Parameters<TPredicate>): Condition<T> } {
    const t = class<T> extends Condition<T> {
        private predicate: Predicate;

        readonly args: Parameters<TPredicate>;

        constructor(...args: Parameters<TPredicate>) {
            super();

            this.args = args;
            this.predicate = predicateFactory(...this.args);
            Object.freeze(this);
        }

        isSatisfied(value: T): Promise<boolean> | boolean {
            return this.predicate(value);
        }

        toString() {
            return descFactory ? descFactory(...this.args) : is.getDescription(this.predicate);
        }

        [NORMALIZE]() {
            return normalizeArgs ? normalizeArgs(...this.args) : this.args;
        }

        static [DENORMALIZE](data: any[]) {
            const newArgs: Parameters<TPredicate> = denormalizeArgs ? denormalizeArgs(...data) : data as Parameters<TPredicate>;
            return new t(...newArgs);
        }
    }

    return t;
}

function caseSensitiveString(isCaseSensitive: boolean) {
    return `(case ${isCaseSensitive ? '' : 'in'}sensitive)`;
}

export namespace Condition {
    export class OnValueExtractor<T> extends Condition<T> {
        constructor(readonly valueExtractor: ValueExtractor<T>, readonly condition: Condition<any>) {
            super();
        }

        async isSatisfied(value: T): Promise<boolean> {
            const extractedValue = await this.valueExtractor.extract(value);
            return this.condition.isSatisfied(extractedValue);
        }

        toString() {
            return `${this.valueExtractor.toString()} ${this.condition.toString()}`;
        }

        [NORMALIZE]() {
            return [this.valueExtractor, this.condition];
        }

        static [DENORMALIZE](data: [any, any]) {
            return new OnValueExtractor(data[0], data[1]);
        }
    }

    abstract class Aggregation<T> extends Condition<T> {
        readonly conditions: Condition[];

        constructor(...conditions: Condition[]) {
            super();
            this.conditions = conditions;
            Object.freeze(this);
        }

        protected conditionsToString() {
            return this.conditions.map(x => x.toString()).join(', ');
        }
    }

    export class All<T> extends Aggregation<T> {
        async isSatisfied(value: any): Promise<boolean> {
            for (const condition of this.conditions) {
                const result = await condition.isSatisfied(value);
                if (result === false) {
                    return false;
                }
            }
            return true;
        }

        toString(): string {
            return `All conditions meet: ${this.conditionsToString()}`;
        }
    }

    export class Any<T> extends Aggregation<T> {
        async isSatisfied(value: T): Promise<boolean> {
            for (const condition of this.conditions) {
                const result = await condition.isSatisfied(value);

                if (result === true) {
                    return true;
                }
            }
            return false;
        }

        toString() {
            return `Any condition meet: ${this.conditionsToString()}`;
        }
    }

    export const Equal = createConditionClass((value: any) => is.eq(value));
    export const NotEqual = createConditionClass(value => is.not(is.eq(value)));
    export const LessThanOrEqual = createConditionClass((value: any) => is.lessThanOrEqual(value));
    export const LessThan = createConditionClass((value: any) => is.lessThan(value));
    export const GreaterThanOrEqual = createConditionClass((value: any) => is.greaterThanOrEqual(value));
    export const GreaterThan = createConditionClass((value: any) => is.greaterThan(value))
    export const StartsWith = createConditionClass(
        (needle: string, isCaseSensitive: boolean = false) => (value: any) => {
            if (!is.string(value)) {
                return false;
            }
            let valueToCompare = value;
            let needleToCompare = needle;
            if (!isCaseSensitive) {
                valueToCompare = valueToCompare.toLowerCase();
                needleToCompare = needleToCompare.toLowerCase();
            }
            return is.startsWith(needleToCompare, valueToCompare);
        },
        (needle, isCaseSensitive = false) => {
            return `a string that starts with "${needle}" ${caseSensitiveString(isCaseSensitive)}`;
        }
    );
    export const EndsWith = createConditionClass(
        (needle: string, isCaseSensitive: boolean = false) => (value: any) => {
            if (!is.string(value)) {
                return false;
            }
            let valueToCompare = value;
            let needleToCompare = needle;
            if (!isCaseSensitive) {
                valueToCompare = valueToCompare.toLowerCase();
                needleToCompare = needleToCompare.toLowerCase();
            }
            return is.endsWith(needleToCompare, valueToCompare);
        },
        (needle, isCaseSensitive = false) => {
            return `a string that ends with "${needle}" ${caseSensitiveString(isCaseSensitive)}`;
        }
    );

    function normalizeRegExp(regexp: RegExp) {
        return [regexp.source, regexp.flags];
    }

    function denormalizeRegexp(source: string, flags: string): [RegExp] {
        return [new RegExp(source, flags)];
    }

    export const Matches = createConditionClass(
        (value: RegExp) => is.matches(value),
        undefined,
        normalizeRegExp,
        denormalizeRegexp
    );
    export const DoesNotMatch = createConditionClass(
        value => is.not(is.matches(value)),
        undefined,
        normalizeRegExp,
        denormalizeRegexp
    );

    export const In = createConditionClass((value: any[]) => is.in(value));
    export const NotIn = createConditionClass((value: any[]) => is.not(is.in(value)));

    export const Contains = createConditionClass(
        (needle: string, isCaseSensitive: boolean = false) => (value: any) => {
            if (!is.string(value)) {
                return false;
            }
            let valueToCompare = value;
            let needleToCompare = needle;
            if (!isCaseSensitive) {
                valueToCompare = valueToCompare.toLowerCase();
                needleToCompare = needleToCompare.toLowerCase();
            }
            return valueToCompare.includes(needleToCompare);
        },
        (needle, isCaseSensitive = false) => `contains "${needle}" ${caseSensitiveString(isCaseSensitive)}`
    );

    export const DoesNotContain = createConditionClass(
        (needle: string, isCaseSensitive: boolean = false) => (value: any) => {
            if (!is.string(value)) {
                return true;
            }
            let valueToCompare = value;
            let needleToCompare = needle;
            if (!isCaseSensitive) {
                valueToCompare = valueToCompare.toLowerCase();
                needleToCompare = needleToCompare.toLowerCase();
            }
            return !valueToCompare.includes(needleToCompare);
        },
        (needle, isCaseSensitive = false) => `does not contain "${needle}" ${caseSensitiveString(isCaseSensitive)}`
    );

    export class IsNull<T> extends Condition<T> {
        isSatisfied(value: T): Promise<boolean> | boolean {
            return is.null(value);
        }

        toString() {
            return 'is null';
        }
    }

    export class NotNull<T> extends Condition<T> {
        isSatisfied(value: T): Promise<boolean> | boolean {
            return is.not(is.null)(value);
        }

        toString() {
            return 'not null';
        }
    }
}