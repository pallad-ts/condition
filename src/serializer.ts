import {Serializer, simpleNormalizer} from "alpha-serializer";
import {ValueExtractor} from "./ValueExtractor";
import {NORMALIZE, DENORMALIZE} from "./global";
import {Condition} from "./Condition";

export function setupSerializer(serializer: Serializer) {

    const normalizer = serializer.normalizer;

    function setup(name: string, clazz: {
        new(...args: any[]): any,
        [DENORMALIZE]?: (value: any) => any
    }) {
        normalizer.registerNormalization({
            name,
            clazz: clazz as any,
            normalizer: (value: any) => {
                if (NORMALIZE in value) {
                    return value[NORMALIZE]();
                }
                return simpleNormalizer(value);
            },
            denormalizer: clazz[DENORMALIZE]
        });
    }

    setup('ValueExtractor/Property', ValueExtractor.Property);

    for (const [name, clazz] of Object.entries(Condition)) {
        if ('isSatisfied' in clazz.prototype) {
            setup(`Condition/${name}`, clazz);
        }
    }
}

