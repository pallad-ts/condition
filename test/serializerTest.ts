import {JSONAdapter, Serializer, StandardNormalizer} from "alpha-serializer";
import {setupSerializer} from "@src/serializer";
import {Condition} from "@src/Condition";
import * as is from "predicates";
import {ValueExtractor} from "@src/ValueExtractor";
import {cases} from "./conditionsTestCases";

describe('serializer', () => {
    let serializer: Serializer;

    beforeEach(() => {
        serializer = new Serializer(
            new JSONAdapter(),
            new StandardNormalizer()
        );

        setupSerializer(serializer);
    });

    describe.each<[string, Condition, any[]]>(
        cases.map(([name, condition, validValues, invalidValues]) => [
            'Condition/' + name,
            condition,
            validValues.concat(invalidValues)
        ])
    )('%s', (name, condition, testValues) => {
        let newCondition: Condition<any>;
        beforeEach(() => {
            newCondition = serializer.deserialize(
                serializer.serialize(condition)
            );
        });

        it('snapshot', () => {
            expect(serializer.serialize(condition))
                .toMatchSnapshot();
        })

        it('has proper serialized name', () => {
            const result = serializer.normalizer.normalize(condition);
            expect(result['@type'])
                .toEqual(name);
        });

        it.each(
            testValues.map(x => [x])
        )('works the same after serialization and deserialization: %s', async value => {
            const result = newCondition.isSatisfied(value);
            const originalResult = condition.isSatisfied(value);
            if (is.promiseLike(result)) {
                return expect(result)
                    .resolves
                    .toEqual(await originalResult);
            }
            expect(result)
                .toEqual(originalResult);
        });

        it('remains the same description', () => {
            expect(newCondition.toString())
                .toEqual(condition.toString());
        })
    });
});