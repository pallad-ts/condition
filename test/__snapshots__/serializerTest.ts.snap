// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializer Condition/All snapshot 1`] = `"{\\"@type\\":\\"Condition/All\\",\\"value\\":[{\\"@type\\":\\"Condition/Contains\\",\\"value\\":[\\"foo\\"]},{\\"@type\\":\\"Condition/Contains\\",\\"value\\":[\\"bar\\"]}]}"`;

exports[`serializer Condition/Any snapshot 1`] = `"{\\"@type\\":\\"Condition/Any\\",\\"value\\":[{\\"@type\\":\\"Condition/Contains\\",\\"value\\":[\\"foo\\"]},{\\"@type\\":\\"Condition/Contains\\",\\"value\\":[\\"bar\\"]}]}"`;

exports[`serializer Condition/Contains snapshot 1`] = `"{\\"@type\\":\\"Condition/Contains\\",\\"value\\":[\\"foo\\"]}"`;

exports[`serializer Condition/Contains snapshot 2`] = `"{\\"@type\\":\\"Condition/Contains\\",\\"value\\":[\\"foo\\",true]}"`;

exports[`serializer Condition/DoesNotContain snapshot 1`] = `"{\\"@type\\":\\"Condition/DoesNotContain\\",\\"value\\":[\\"foo\\"]}"`;

exports[`serializer Condition/DoesNotContain snapshot 2`] = `"{\\"@type\\":\\"Condition/DoesNotContain\\",\\"value\\":[\\"foo\\",true]}"`;

exports[`serializer Condition/DoesNotMatch snapshot 1`] = `"{\\"@type\\":\\"Condition/DoesNotMatch\\",\\"value\\":[\\"a{1,2}\\",\\"\\"]}"`;

exports[`serializer Condition/EndsWith snapshot 1`] = `"{\\"@type\\":\\"Condition/EndsWith\\",\\"value\\":[\\"foo\\"]}"`;

exports[`serializer Condition/EndsWith snapshot 2`] = `"{\\"@type\\":\\"Condition/EndsWith\\",\\"value\\":[\\"foo\\",true]}"`;

exports[`serializer Condition/Equal snapshot 1`] = `"{\\"@type\\":\\"Condition/Equal\\",\\"value\\":[10]}"`;

exports[`serializer Condition/GreaterThan snapshot 1`] = `"{\\"@type\\":\\"Condition/GreaterThan\\",\\"value\\":[10]}"`;

exports[`serializer Condition/GreaterThanOrEqual snapshot 1`] = `"{\\"@type\\":\\"Condition/GreaterThanOrEqual\\",\\"value\\":[10]}"`;

exports[`serializer Condition/In snapshot 1`] = `"{\\"@type\\":\\"Condition/In\\",\\"value\\":[[1,2]]}"`;

exports[`serializer Condition/IsNull snapshot 1`] = `"{\\"@type\\":\\"Condition/IsNull\\",\\"value\\":{}}"`;

exports[`serializer Condition/LessThan snapshot 1`] = `"{\\"@type\\":\\"Condition/LessThan\\",\\"value\\":[10]}"`;

exports[`serializer Condition/LessThanOrEqual snapshot 1`] = `"{\\"@type\\":\\"Condition/LessThanOrEqual\\",\\"value\\":[10]}"`;

exports[`serializer Condition/Matches snapshot 1`] = `"{\\"@type\\":\\"Condition/Matches\\",\\"value\\":[\\"a{1,2}\\",\\"\\"]}"`;

exports[`serializer Condition/NotEqual snapshot 1`] = `"{\\"@type\\":\\"Condition/NotEqual\\",\\"value\\":[10]}"`;

exports[`serializer Condition/NotIn snapshot 1`] = `"{\\"@type\\":\\"Condition/NotIn\\",\\"value\\":[[1,2]]}"`;

exports[`serializer Condition/NotNull snapshot 1`] = `"{\\"@type\\":\\"Condition/NotNull\\",\\"value\\":{}}"`;

exports[`serializer Condition/OnValueExtractor snapshot 1`] = `"{\\"@type\\":\\"Condition/OnValueExtractor\\",\\"value\\":[{\\"@type\\":\\"ValueExtractor/Property\\",\\"value\\":[\\"test\\"]},{\\"@type\\":\\"Condition/Equal\\",\\"value\\":[10]}]}"`;

exports[`serializer Condition/StartsWith snapshot 1`] = `"{\\"@type\\":\\"Condition/StartsWith\\",\\"value\\":[\\"foo\\"]}"`;

exports[`serializer Condition/StartsWith snapshot 2`] = `"{\\"@type\\":\\"Condition/StartsWith\\",\\"value\\":[\\"foo\\",true]}"`;
